src/erc721/ERC721.sol

```diff
@@ -0,0 +1,193 @@
+        
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    function transfer(address to, uint256 tokenId) external {
+        require(msg.sender == ownerOf[tokenId], "NOT_OWNER");
+        
+        balanceOf[msg.sender]--; 
+        
+        balanceOf[to]++; 
+        
+        getApproved[tokenId] = address(0);
+        
+        ownerOf[tokenId] = to;
+        
+        emit Transfer(msg.sender, to, tokenId); 
+    }
+    
+    function transferFrom(address, address to, uint256 tokenId) external {
+        address owner = ownerOf[tokenId];
+        
+        require(
+            msg.sender == owner || getApproved[tokenId] == msg.sender 
+            || isApprovedForAll[owner][msg.sender], "NOT_AUTHORIZED"
+        );
+        
+        balanceOf[owner]--; 
+        
+        balanceOf[to]++;
+        
+        getApproved[tokenId] = address(0);
+        
+        ownerOf[tokenId] = to;
+        
+        emit Transfer(owner, to, tokenId); 
+    }
+    
+    /*///////////////////////////////////////////////////////////////
+                          PERMIT/EIP-2612 LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    function permit(
+        address owner,
+        address spender,
+        uint256 tokenId,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external {
+        require(deadline >= block.timestamp, "PERMIT_DEADLINE_EXPIRED");
+        // This is reasonably safe because incrementing past type(uint256).max
+        // through approvals is exceedingly unlikely!
+        unchecked {
+            bytes32 digest = keccak256(
+                abi.encodePacked(
+                    "\x19\x01",
+                    DOMAIN_SEPARATOR,
+                    keccak256(abi.encode(PERMIT_TYPEHASH, owner, spender, tokenId, nonces[owner]++, deadline))
+                )
+            );
+
+            address recoveredAddress = ecrecover(digest, v, r, s);
+            require(recoveredAddress != address(0) && recoveredAddress == owner, "INVALID_PERMIT_SIGNATURE");
+            require(recoveredAddress == ownerOf[tokenId], "INVALID_OWNER");
+        }
+
+        getApproved[tokenId] = spender;
+
+        emit Approval(owner, spender, tokenId);
+    }
+    
+    /*///////////////////////////////////////////////////////////////
+                          INTERNAL UTILS
+    //////////////////////////////////////////////////////////////*/
+    
+    function mint(address to, uint256 tokenId, string calldata _tokenURI) internal { 
+        require(ownerOf[tokenId] == address(0), "TOKEN_ID_NOT_UNIQUE");
+        
+        totalSupply++;
+        
+        balanceOf[to]++;
+        
+        ownerOf[tokenId] = to;
+        
+        tokenURI[tokenId] = _tokenURI;
+        
+        emit Transfer(address(0), to, tokenId); 
+    }
+    
+    function burn(address from, uint256 tokenId) internal { 
+        require(ownerOf[tokenId] == from, "NOT_FROM_OWNER");
```

Comment:
whats the point of accepting `from` here at all then if we do this storage load?

Reply:
yeah, I guess this safety check doesn't do that much at base level -- can just remove the storage load, as well. The idea here though, is basically ... want predictability, so we don't have Alice as `from` param and end up burning a token Alice doesn't actually own. But, I can see how this can be opinionated, and controls can be put into inheritance.